# Analisis y Limpieza de datos con Pandas
Voy a dividir la parte del analisis de la limpieza

Como vamos a trabajar con la libreria pandas lo primero que hacemos es importarla

In [4]:
import pandas as pd

#### Leemos los datasets
Para un mayor orden a todos los datasets los puse dentro de la carpeta datasets

In [5]:
df_ventas = pd.read_excel('datasets/datasets_originales/ventas.xlsx')
df_detalle = pd.read_excel('datasets/datasets_originales//detalle_ventas.xlsx')
df_clientes = pd.read_excel('datasets/datasets_originales//clientes.xlsx')
df_productos = pd.read_excel('datasets/datasets_originales//productos.xlsx')


# Etapa analisis
Empezamos explorando los dataframes

Lo hare mediante una funcion para poder reutilizar en cada dataframe

In [6]:
# vamos a hacer una funcion para inspeccionar los dataframes
def inspeccionar_df(df, nombre_df):
    print(f"Inspeccionando el DataFrame: {nombre_df}")
    print("\nfilas, columnas:")
    print(df.shape)
    print("\nPrimeras filas:")
    print(df.head())
    print("\nInformación del DataFrame:")
    print(df.info())
    print("\nDescripción estadística:")
    print(df.describe(include='all'))
    print("\nValores nulos por columna:")
    print(df.isnull().sum())
    print("\nFilas duplicadas:", df.duplicated().sum())
    
    

Ahora podemos analizar uno por uno cada dataframe solo llamando a la funcion inspeccionar_df

## Dataframe clientes

In [7]:
inspeccionar_df(df_clientes, 'df_clientes')

Inspeccionando el DataFrame: df_clientes

filas, columnas:
(100, 5)

Primeras filas:
   id_cliente   nombre_cliente                     email      ciudad  \
0           1    Mariana Lopez    mariana.lopez@mail.com  Carlos Paz   
1           2    Nicolas Rojas    nicolas.rojas@mail.com  Carlos Paz   
2           3  Hernan Martinez  hernan.martinez@mail.com  Rio Cuarto   
3           4     Uma Martinez     uma.martinez@mail.com  Carlos Paz   
4           5  Agustina Flores  agustina.flores@mail.com     Cordoba   

  fecha_alta  
0 2023-01-01  
1 2023-01-02  
2 2023-01-03  
3 2023-01-04  
4 2023-01-05  

Información del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   id_cliente      100 non-null    int64         
 1   nombre_cliente  100 non-null    object        
 2   email           100 non-null    object        
 

Luego de este analisis podemos concluir que:

- Tenemos 100 filas y 5 columnas 

- Los nombres de la columnas son correctos (minusculas y separados por guion bajo)

- No hay valores nulos

- Los tipos de datos son correctos



- Moda:
        
    - *nombre_cliente*: Bruno Castro (2)

    - *Ciudad*: Rio Cuarto (23)

- No hay filas duplicadas

## Dataframe productos

In [8]:
inspeccionar_df(df_productos, 'df_productos')

Inspeccionando el DataFrame: df_productos

filas, columnas:
(100, 4)

Primeras filas:
   id_producto     nombre_producto  categoria  precio_unitario
0            1      Coca Cola 1.5L  Alimentos             2347
1            2          Pepsi 1.5L   Limpieza             4973
2            3         Sprite 1.5L  Alimentos             4964
3            4  Fanta Naranja 1.5L   Limpieza             2033
4            5  Agua Mineral 500ml  Alimentos             4777

Información del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   id_producto      100 non-null    int64 
 1   nombre_producto  100 non-null    object
 2   categoria        100 non-null    object
 3   precio_unitario  100 non-null    int64 
dtypes: int64(2), object(2)
memory usage: 3.3+ KB
None

Descripción estadística:
        id_producto nombre_producto  categoria  precio

Luego de este analisis podemos concluir que:

- Tenemos 100 filas y 4 columnas 

- Los nombres de la columnas son correctos (minusculas y separados por guion bajo)

- No hay valores nulos

- ❌**Hay productos mal categorizados**

- ❌**precio_unitario tiene es entero y deberia ser decimal**

- Solo tenemos 2 categorias de productos

- No hay moda, ya que todos tienen igual frecuencia tanto en **nombre_producto** como **categoria**

- Precio_unitario:
    - **Promedio** = $2718.55
    - **Mediana** = $2516.00

- No hay filas duplicadas

## Dataframe ventas

In [9]:
inspeccionar_df(df_ventas, 'df_ventas')

Inspeccionando el DataFrame: df_ventas

filas, columnas:
(120, 6)

Primeras filas:
   id_venta      fecha  id_cliente    nombre_cliente  \
0         1 2024-06-19          62  Guadalupe Romero   
1         2 2024-03-17          49      Olivia Gomez   
2         3 2024-01-13          20      Tomas Acosta   
3         4 2024-02-27          36    Martina Molina   
4         5 2024-06-11          56        Bruno Diaz   

                       email     medio_pago  
0  guadalupe.romero@mail.com        tarjeta  
1      olivia.gomez@mail.com             qr  
2      tomas.acosta@mail.com        tarjeta  
3    martina.molina@mail.com  transferencia  
4        bruno.diaz@mail.com        tarjeta  

Información del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120 entries, 0 to 119
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   id_venta        120 non-null    int64         
 1   fecha      

Luego de este analisis podemos concluir que:

- Tenemos 120 filas y 6 columnas 

- Los nombres de la columnas son correctos (minusculas y separados por guion bajo)

- No hay valores nulos

- Los tipos de datos son correctos

- Hay 67 cuentas de clientes diferentes (email)

- Hay 4 medios de pagos

- Moda:
    
    - *nombre_cliente*: Bruno Diaz (5)

    - *email*: bruno.diaz@mail.com (5)

- No hay filas duplicadas

## Dataframe detalle_ventas

In [10]:
inspeccionar_df(df_detalle, 'df_detalle')

Inspeccionando el DataFrame: df_detalle

filas, columnas:
(343, 6)

Primeras filas:
   id_venta  id_producto        nombre_producto  cantidad  precio_unitario  \
0         1           90    Toallas Húmedas x50         1             2902   
1         2           82  Aceitunas Negras 200g         5             2394   
2         2           39     Helado Vainilla 1L         5              469   
3         2           70           Fernet 750ml         2             4061   
4         2           22  Medialunas de Manteca         1             2069   

   importe  
0     2902  
1    11970  
2     2345  
3     8122  
4     2069  

Información del DataFrame:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 343 entries, 0 to 342
Data columns (total 6 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   id_venta         343 non-null    int64 
 1   id_producto      343 non-null    int64 
 2   nombre_producto  343 non-null    object
 3   cantida

Luego de este analisis podemos concluir que:

- Tenemos 343 filas y 6 columnas 

- Los nombres de la columnas son correctos (minusculas y separados por guion bajo)

- No hay valores nulos

- ❌**Los tipos de datos precio_unitario e importe deberian ser del tipo decimal (estan como enteros)** 

- Hay 95 productos diferentes (nombre_producto)

- Moda: *nombre_producto*: Queso Rallado 150g (9)

- No hay filas duplicadas

----
Ya analice todos los dataframes y los casos a tener en cuenta a la hora de limpiar los datos estan en negrita y con una cruz roja (❌)

---

# Etapa de Limpieza
Para mantener los archivos originales y poder comparar el antes y el despues, voy a hacer una copia de todos los dataframes y modificar las copias, a estos nuevos datasets los pondré en una subcarpeta llamada datasets_limpios

In [11]:
df_clientes_limpio = df_clientes.copy()
df_productos_limpio = df_productos.copy()
df_ventas_limpio = df_ventas.copy()
df_detalle_limpio = df_detalle.copy()

# ahora guardamos los datasets en la subcarpeta datasets_limpios

df_clientes_limpio.to_excel('datasets/datasets_limpios/clientes_limpio.xlsx', index=False)
df_productos_limpio.to_excel('datasets/datasets_limpios/productos_limpio.xlsx', index=False)
df_ventas_limpio.to_excel('datasets/datasets_limpios/ventas_limpio.xlsx', index=False)
df_detalle_limpio.to_excel('datasets/datasets_limpios/detalle_ventas_limpio.xlsx', index=False)

Ahora si podemos empezar a hacer la limpieza a las copias de los dataframes

# Limpieza general

Primero normalizaremos los nombres de las columnas, sacando los espacios del inicio y del final y convirtiendo todo a minusculas con un guion bajo como separador, en vez de un espacio, ya de esta forma es la que se trabaja normalmente y evita errores, hago esto por mas que no lo haya detectado en la etapa del analisis porque es un error comun.

Para eso hacemos una funcion que normalice las columnas

In [12]:
def normalizar_nombres_columnas(df):
    df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')
    return df

Ahora podemos aplicar la funcion a cada dataframe

In [13]:
df_clientes_limpio = normalizar_nombres_columnas(df_clientes_limpio)
df_productos_limpio = normalizar_nombres_columnas(df_productos_limpio)
df_ventas_limpio = normalizar_nombres_columnas(df_ventas_limpio)
df_detalle_limpio = normalizar_nombres_columnas(df_detalle_limpio)

# Dataframe clientes

Comenzamos viendo si hay nombres de clientes duplicados

In [14]:
# Identificar si hay nombres de clientes duplicados
df_clientes_limpio[df_clientes_limpio.duplicated(subset=['nombre_cliente'], keep=False)]

id_cliente     nombre_cliente                        email       ciudad  \
7            8       Bruno Castro        bruno.castro@mail.com   Carlos Paz   
9           10      Karina Acosta       karina.acosta@mail.com      Cordoba   
20          21    Elena Rodriguez     elena.rodriguez@mail.com  Alta Gracia   
33          34       Bruno Castro       bruno.castro2@mail.com  Villa Maria   
40          41    Elena Rodriguez    elena.rodriguez2@mail.com  Alta Gracia   
45          46  Agustina Martinez   agustina.martinez@mail.com  Alta Gracia   
57          58      Karina Acosta      karina.acosta2@mail.com   Rio Cuarto   
78          79       Olivia Perez        olivia.perez@mail.com   Carlos Paz   
84          85  Agustina Martinez  agustina.martinez2@mail.com   Mendiolaza   
94          95       Olivia Perez       olivia.perez2@mail.com   Rio Cuarto   

   fecha_alta  
7  2023-01-08  
9  2023-01-10  
20 2023-01-21  
33 2023-02-03  
40 2023-02-10  
45 2023-02-15  
57 2023-02-27  
78 2023-03-20  
84 2023-03-26  
94 2023-04-05

Hay nombres que se repiten, pero pueden tratarse de personas diferentes con el mismo nombre, ya que la mayoria viven en ciudades diferentes y tienen un email diferente, por lo tanto podemos tomarlo como que no es un error

Corregidos los errores guardo el dataframe

In [15]:
df_clientes_limpio.to_excel('datasets/datasets_limpios/clientes_limpio.xlsx', index=False)

# Dataframe productos
En el analisis habiamos detectado lo siguiente:
- Habia productos mal categorizados
- precio_unitario debe ser decimal

Empezamos poniendo el tipo de dato correspondiente a precio_unitario

In [16]:
#pasar de entero a decimal la columna precio_unitario
df_productos_limpio['precio_unitario'] = df_productos_limpio['precio_unitario'].astype(float)
print("Se cambio el tipo de dato a precio_unitario, ahora el tipo de dato es:", df_productos_limpio['precio_unitario'].dtypes)

Se cambio el tipo de dato a precio_unitario, ahora el tipo de dato es: float64


Ahora vamos a categorizar bien los productos, primero vamos a ver que categorias tenemos

In [17]:
print(df_productos_limpio['categoria'].unique())
print(df_productos_limpio['categoria'].value_counts())

['Alimentos' 'Limpieza']
categoria
Alimentos    50
Limpieza     50
Name: count, dtype: int64


Tenemos unicamente 2 categorias "Alimentos" y "Limpieza", ambas con la misma cantidad de productos, lo ideal seria hacerlo con machine learning, pero por el momento lo haremos manual

In [18]:
#quiero categorizar los productos de forma correcta, los alimentos
#Los productos de limpieza son los que tienen estas palabras: lavandina, shampoo, jabon, detergente, papel, servilleta, crema /hilo dental, suaviznte, desengrasante, esponja, trapo, toallas, desodorante, cepillo, mascarilla, limpiavidrios
#los demas son alimentos
categoria_limpieza = ['lavandina', 'shampoo', 'jabon', 'tocador', 'detergente', 'papel', 'servilleta', 'crema dental', 'hilo dental', 'suavizante', 'desengrasante', 'esponja', 'trapo', 'toallas', 'desodorante', 'cepillo', 'mascarilla', 'limpiavidrios']
def categorizar_producto(nombre_producto):
    nombre_producto = nombre_producto.lower()
    for palabra in categoria_limpieza:
        if palabra in nombre_producto:
            return 'Limpieza'
    return 'Alimentos'
df_productos_limpio['categoria'] = df_productos_limpio['nombre_producto'].apply(categorizar_producto)
print(df_productos_limpio['categoria'].value_counts())



categoria
Alimentos    83
Limpieza     17
Name: count, dtype: int64


Vemos que la mayoria de los productos eran alimentos y muchos productos estaban mal categorizados

Puedo ver si hay valores outliers

Hare una funcion para ver si hay valores outliers

In [19]:
#funcion que me dice si hay valores outliers en una columna numerica
def detectar_outliers(df, columna):
    Q1 = df[columna].quantile(0.25)
    Q3 = df[columna].quantile(0.75)
    IQR = Q3 - Q1
    limite_inferior = Q1 - 1.5 * IQR
    limite_superior = Q3 + 1.5 * IQR
    outliers = df[(df[columna] < limite_inferior) | (df[columna] > limite_superior)]
    return outliers



In [20]:
outliers_precio = detectar_outliers(df_productos_limpio, 'precio_unitario')
print("Outliers en la columna precio_unitario:")

Outliers en la columna precio_unitario:


No hay outliers

Corregidos los errores guardo el dataframe

In [21]:
df_productos_limpio.to_excel('datasets/datasets_limpios/productos_limpio.xlsx', index=False)

# Dataframe ventas

vemos cuantas categorias hay en medio de pago, y si hay algun dato mal cargado

In [22]:
df_ventas_limpio["medio_pago"].unique()

array(['tarjeta', 'qr', 'transferencia', 'efectivo'], dtype=object)

No se encontraron errores en los datos, puedo guardarlo al dataframe, pero no se le hizo ningun cambio

In [23]:
df_ventas_limpio.to_excel('datasets/datasets_limpios/ventas_limpio.xlsx', index=False)

# Dataframe detalle_ventas
En el analisis habiamos detectado lo siguiente:
- precio_unitario debe ser decimal
- importe debe ser decimal


In [24]:
df_detalle_limpio["precio_unitario"] = df_detalle_limpio["precio_unitario"].astype(float)
df_detalle_limpio["importe"] = df_detalle_limpio["importe"].astype(float)
print("Se cambio el tipo de dato a precio_unitario, ahora el tipo de dato es:", df_detalle_limpio['precio_unitario'].dtypes)
print("Se cambio el tipo de dato a importe, ahora el tipo de dato es:", df_detalle_limpio['importe'].dtypes)


Se cambio el tipo de dato a precio_unitario, ahora el tipo de dato es: float64
Se cambio el tipo de dato a importe, ahora el tipo de dato es: float64


Si bien no tenemos lineas duplicadas, observe que hay productos identicos que se repiten dentro de una misma venta, cuando este deberia estar solo una vez y variar la cantidad

In [25]:
detalle_unificado = (
    df_detalle_limpio
    .groupby(['id_venta', 'id_producto', 'nombre_producto'], as_index=False)
    .agg({
        'cantidad': 'sum',
        'precio_unitario': 'first',  # asumimos que el precio es el mismo
        'importe': 'sum'
    })
)
#agrupo por id_venta, id_producto y nombre_producto, sumo la cantidad y el importe, y tomo el primer precio_unitario ya que asumimos que el precio es el mismo para un mismo producto en una misma venta
#esto es para evitar tener el mismo producto repetido en una misma venta

#veremos la longitud del dataframe antes y despues de unificar para ver la diferencia
print("Antes:", len(df_detalle_limpio))
print("Después:", len(detalle_unificado))



Antes: 343
Después: 339


como vemos cambio la longitud del dataset,
se redujeron 4 filas, esto quiere decir que se unificaron productos, entonces debo asignarlo al dataframe limpio

In [26]:
df_detalle_limpio = detalle_unificado

Resuelto todos los problemas podemos guardar el dataframe

In [27]:
df_detalle_limpio.to_excel('datasets/datasets_limpios/detalle_ventas_limpio.xlsx', index=False)

Ahora puedo validar que todos los id_cliente en ventas existan en clientes

In [28]:
df_ventas_limpio['id_cliente'].isin(df_clientes_limpio['id_cliente']).all()
#esto devuelve True si todos los id_cliente en ventas existen en clientes

np.True_

Vemos que todos los clientes que hicieron una compra existen y son validos

Ahora haremos lo mismo con los productos que estan en el detalle de venta

In [29]:
df_detalle_limpio['id_producto'].isin(df_productos_limpio['id_producto']).all()
#esto devuelve True si todos los id_producto en detalle de ventas existen en productos

np.True_

Esto muestra que no hay detalle de ventas asociadas a productos que no esten en el dataset productos

# Analisis estadistico descriptivo
Si bien ya hicimos un analisis , ahora vamos a analizar las estadisticas con los dataframes ya limpios

### Dataframe Clientes 
Podemos analizar cual es la ciudad de donde estan la mayoria de nuestros Clientes

No podemos analizar tipo de distribucion ya que no contamos con columnas numericas ademas de los identificadores

In [73]:
df_clientes_limpio["ciudad"].mode()[0]

'Rio Cuarto'

Rio Cuarto es la ciudad que más clientes nos aporta

Correlacion en clientes no tiene sentido ya que no hay columnas numericas

### Dataframe productos

In [49]:
print(df_productos_limpio["precio_unitario"].describe(include='all').round(2))

count     100.00
mean     2718.55
std      1381.64
min       272.00
25%      1590.00
50%      2516.00
75%      4026.50
max      4982.00
Name: precio_unitario, dtype: float64


- Estadisticas basicas de precio_unitario
    
    - Media: 2718.55

    - Mediana: 2516

    - Moda: No tiene

    - Desviacion:1381.64

    - Rango: [272,4982]
    
- La media es mayor que la mediana, lo cual indica que la distribución está sesgada hacia la derecha (sesgo positivo).
Esto significa que hay algunos productos con precios altos que elevan el promedio general, aunque la mayoría de los productos tienen precios más bajos.

- La desviación estándar elevada refleja una alta variabilidad en los precios de los productos, lo que sugiere la presencia de diferentes gamas o tipos de productos dentro del catálogo.

La correlacion no tiene sentido para productos

### Dataframe ventas

Lo interesante para analizar en este dataframe son las modas de medio de pago y nombre_cliente para saber quien es el que mas compro y cual es el medio de pago mas utilizado

No podemos analizar tipo de distribucion ya que no contamos con columnas numericas ademas de los identificadores

In [ ]:
# Moda de la columna medio_pago
print(df_ventas_limpio['medio_pago'].mode()[0])
# Moda de nombre_cliente
print(df_ventas_limpio['nombre_cliente'].mode()[0])


efectivo
Bruno Diaz


El que mas compro fue Bruno Diaz y el medio de pago mas utilizado es el efectivo

La correlacion no tiene sentido en ventas

### Dataframe detalle de ventas

Lo interesante a analizar es cual es el producto presente en mas ventas 

In [71]:
# Moda de nombre_producto
print(df_detalle_limpio['nombre_producto'].mode()[0])

Queso Rallado 150g


El producto de moda es el queso rallado, es el producto que mas gente se llevo, sin tener en cuenta la cantidad

In [72]:
# quiero saber cual es el producto que mas se vende en cantidad
producto_mas_vendido = df_detalle_limpio.groupby('nombre_producto')['cantidad'].sum().idxmax()
print("El producto que mas se vende en cantidad es:", producto_mas_vendido)

El producto que mas se vende en cantidad es: Salsa de Tomate 500g


El producto que mas se vendió teniendo en cuenta la cantidad es la salsa de Tomate

Ahora puedo analizar las estadisticas de las columna cantidad

In [79]:
df_detalle_limpio["cantidad"].describe(include='all').round(2)

count    339.00
mean       3.00
std        1.43
min        1.00
25%        2.00
50%        3.00
75%        4.00
max        8.00
Name: cantidad, dtype: float64

- Estadisticas basicas de cantidad
    
    - Media: 3

    - Mediana: 3

    - Moda: No tiene

    - Desviacion: 1.43

    - Rango: [1,8]

- La Media y la Mediana son iguales, por lo que quiere decir que la distribucion es normal o simetrica

- La desviacion es baja lo que refleja poca variabilidad

In [82]:
#correlacion entre cantidad y precio_unitario en df_detalle_limpio
correlacion = df_detalle_limpio['cantidad'].corr(df_detalle_limpio['precio_unitario'])
print("La correlacion entre cantidad y precio_unitario es:", correlacion)


La correlacion entre cantidad y precio_unitario es: -0.0599226518423504


La correlación entre cantidad y precio_unitario es –0.06, lo que indica una relación prácticamente nula entre ambas variables.
En otras palabras, los cambios en el precio unitario no parecen afectar la cantidad vendida de manera lineal.